### PR TITLE
Copy Headers

### PR DESCRIPTION
**Pull request by RichardHightower**
# Copy Headers.
We defined a binary protocol to encode headers into a Nat's message and this allows us to pass headers from JMS through nats and back. The header copy is bidirectional. 

* #38 JMS Bridge header copy test
* #34 Got rid of String Message 
* Started ..4 and created VMs but not done, but made JNDI configurable and URL are now env vars
* #35 Defined Message Protocol for encoding headers
* ..36 Started and did some task to facilitate this (progress)
* #37 Did this to support 4 and 36 

To see headers being copied from Nats to JMS see `NatsToJmsBridgeWithHeadersTest`.
To see headers being copied from JMS to Nats see `JmsToNatsBridgeWithHeadersTest`.

You can pass these env vars if you want to run tests against, for example, docker containers or kube instance. 

```
NATS_BRIDGE_JMS_CONNECTION_FACTORY = tcp://localhost:61616
NATS_BRIDGE_NATS_SERVERS  = nats://localhost:4222
```
### Current Master 
```
Total commits: 14
Total ctimes: 185
Total files: 30
Total loc: 1839
| Author            |   loc |   coms |   fils |  distribution    |
|:------------------|------:|-------:|-------:|:-----------------|
| Richard Hightower |  1839 |     14 |     30 | 100.0/ 100/100.0 |

```
#### This branch 
```
Total commits: 31
Total ctimes: 869
Total files: 102
Total loc: 5253
| Author            |   loc |   coms |   fils |  distribution   |
|:------------------|------:|-------:|-------:|:----------------|
| Richard Hightower |  3621 |     26 |     64 | 68.9/83.9/62.7  |
| Rick Hightower    |  1632 |      5 |     38 | 31.1/16.1/37.3  |

```
* 286% more code than before. Most of that is copying headers and implementing the wire protocol. 
* We are at 90% code coverage. 
* There are full integration tests that send headers from Nats, use those headers behind JMS/MQ and then send them back. 
* All integration tests are mirrored, e.g., full integration tests that send headers from JMS, use those headers behind Nats and then send them back. 
* Request reply bridge is fully implemented 
* Queue to queue bridge is fully implemented 
* Passing JMS/MQ headers are fully implemented 

### VERSION AND MARKER

* `<AB marker [byte]>`
* `<CD marker [byte]>`
* `<MAJOR_VERSION [byte]>`
* `<MINOR_VERSION[byte]>`
* `<AB marker [byte]>`
* `<CD marker [byte]>`

###  HEADERS

* `<HEADER_LEN [int]>`
* `<HEADER_HASH [int]>`
* `<HEADER_JSON_OBJECT_BYTES [bytes]>`

###  OPAQUE BODY
* `<BODY_LEN [int]>`
* `<BODY_HASH [int]>`
* `<OPAQUE_BODY_BYTES>[bytes]`

#### Protocol.java
```java

public class Protocol {

    public static final int MESSAGE_VERSION_MAJOR = 1;
    public static final int MESSAGE_VERSION_MINOR = 0;
    public static final int MARKER_AB = (byte) 0xab;
    public static final int MARKER_CD = (byte) 0xcd;

    public static final String HEADER_KEY_DELIVERY_TIME = "BRIDGE_deliveryTime";
    public static final String HEADER_KEY_TIMESTAMP = "BRIDGE_timestamp";
    public static final String HEADER_KEY_MODE = "BRIDGE_mode";
    public static final String HEADER_KEY_EXPIRATION_TIME = "BRIDGE_expirationTime";
    public static final String HEADER_KEY_TYPE = "BRIDGE_type";
    public static final String HEADER_KEY_PRIORITY = "BRIDGE_priority";
    public static final String HEADER_KEY_REDELIVERED = "BRIDGE_redelivered";


    public static int createHashCode(byte[] value) {
        int h = 0;
        byte[] var2 = value;
        int var3 = value.length;

        for (int var4 = 0; var4 < var3; ++var4) {
            byte v = var2[var4];
            h = 31 * h + (v & 255);
        }

        return h;
    }

}

```

#### Bridge Message interface that defines common headers

```java
package io.nats.bridge.messages;

import java.nio.charset.StandardCharsets;
import java.util.Collections;
import java.util.Map;

public interface Message {


    String NO_TYPE = "NO_TYPE";

    default long timestamp() {return -1L;}
    //TTL plus timestamp
    default long expirationTime() {return -1L;}
    //Delivery time is not instant
    default long deliveryTime() {return -1L;}
    default int deliveryMode() {return -1;}
    default String type() {return NO_TYPE;}
    default boolean redelivered() {return false;}
    default int priority() {return -1;}

    default String correlationID() {return "";}

    default Map<String, Object> headers() {return Collections.emptyMap();}

    default void reply(Message reply) {}

    default byte [] getBodyBytes() {
        return new byte[0];
    }

    default String bodyAsString() {
        return new String(getBodyBytes(), StandardCharsets.UTF_8);
    }

    default byte [] getMessageBytes() {
        return getBodyBytes();
    }
}

```

#### BaseMessageWithHeaders.java - Encoding headers in byte array 

 ```java 
package io.nats.bridge.messages;
...
public class BaseMessageWithHeaders implements BytesMessage {

...
    public byte[] getMessageAsBytes() {

        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
        final DataOutputStream streamOut = new DataOutputStream(baos);

        try {


            streamOut.writeByte(Protocol.MARKER_AB);
            streamOut.writeByte(Protocol.MARKER_CD);
            streamOut.writeByte(Protocol.MESSAGE_VERSION_MAJOR);
            streamOut.writeByte(Protocol.MESSAGE_VERSION_MINOR);
            streamOut.writeByte(Protocol.MARKER_AB);
            streamOut.writeByte(Protocol.MARKER_CD);


            final HashMap<String, Object> outputHeaders = new HashMap<>(9 + (headers != null ? headers.size() : 0));

            if (headers != null) {
                outputHeaders.putAll(headers);
            }
            if (deliveryTime > 0) {
                outputHeaders.put(HEADER_KEY_DELIVERY_TIME, this.deliveryTime());
            }
            if (mode != -1)
                outputHeaders.put(HEADER_KEY_MODE, this.deliveryMode());
            if (expirationTime > 0)
                outputHeaders.put(HEADER_KEY_EXPIRATION_TIME, this.expirationTime());
            if (timestamp > 0)
                outputHeaders.put(HEADER_KEY_TIMESTAMP, this.timestamp());
            if (type != null)
                outputHeaders.put(HEADER_KEY_TYPE, this.type());
            if (priority != -1)
                outputHeaders.put(HEADER_KEY_PRIORITY, this.priority());
            if (redelivered)
                outputHeaders.put(HEADER_KEY_REDELIVERED, this.redelivered());

            byte[] headerBytes = mapper.writeValueAsBytes(outputHeaders);
            streamOut.writeInt(headerBytes.length);
            streamOut.writeInt(Protocol.createHashCode(headerBytes));
            streamOut.write(headerBytes);

            //ystem.out.println("header bytes length " + headerBytes.length);
            //ystem.out.println("header bytes hash " + Protocol.createHashCode(headerBytes));


            if (bodyBytes != null) {


                streamOut.writeInt(bodyBytes.length);
                streamOut.writeInt(Protocol.createHashCode(bodyBytes));
                //ystem.out.println("body bytes length " + bodyBytes.length);
                //ystem.out.println("body bytes hash " + Protocol.createHashCode(bodyBytes));
                streamOut.write(bodyBytes);
            } else {
                streamOut.write(0);
                streamOut.write(0);
            }


        } catch (Exception e) {
            throw new MessageException("Can't write out message", e);
        } finally {
            try {
                streamOut.close();
                baos.close();
            } catch (Exception e) {
                throw new MessageException("Can't write out message", e);
            }
        }
        return baos.toByteArray();
    }
...
}
```

#### MessageBuilder - turn bytes into message 
```java

public class MessageBuilder {

    public Message buildFromBytes(byte[] buffer) {

        if (buffer.length > 5) {

            if (buffer[0] == MARKER_AB &&
                    buffer[1] == MARKER_CD &&
                    buffer[2] == Protocol.MESSAGE_VERSION_MAJOR &&
                    buffer[3] == Protocol.MESSAGE_VERSION_MINOR &&
                    buffer[4] == MARKER_AB &&
                    buffer[5] == MARKER_CD
            ) {

                final DataInputStream dataInputStream = new DataInputStream(new ByteArrayInputStream(buffer));
                try {
                    dataInputStream.skipBytes(6);

                    //Read the header
                    final int jsonLength = dataInputStream.readInt();
                    final int jsonHash = dataInputStream.readInt();
                    final byte[] jsonByteBuffer = new byte[jsonLength];
                    dataInputStream.read(jsonByteBuffer);
                    if (Protocol.createHashCode(jsonByteBuffer) != jsonHash) {
                        throw new MessageBuilderException("JSON Hash did not match for headers");
                    }
                    final Map<String, Object> header = mapper.readValue(jsonByteBuffer, Map.class);

                    //Read the body
                    final int bodyLength = dataInputStream.readInt();

                    //ystem.out.println("body len " + bodyLength);
                    final int bodyHash = dataInputStream.readInt();
                    //ystem.out.println("body hash " + bodyHash);
                    final byte[] bodyBuffer = new byte[bodyLength];
                    dataInputStream.read(bodyBuffer);
                    if (Protocol.createHashCode(bodyBuffer) != bodyHash) {
                        throw new MessageBuilderException("Body Hash did not match ");
                    }

                    /* read headers */
                    if (header.containsKey(HEADER_KEY_TIMESTAMP)) {
                        withTimestamp((long) header.get(HEADER_KEY_TIMESTAMP));
                        header.remove(HEADER_KEY_TIMESTAMP);
                    }
                    if (header.containsKey(HEADER_KEY_EXPIRATION_TIME)) {
                        withExpirationTime((long) header.get(HEADER_KEY_EXPIRATION_TIME));
                        header.remove(HEADER_KEY_EXPIRATION_TIME);
                    }
                    if (header.containsKey(HEADER_KEY_DELIVERY_TIME)) {
                        withDeliveryTime((long) header.get(HEADER_KEY_DELIVERY_TIME));
                        header.remove(HEADER_KEY_DELIVERY_TIME);
                    }
                    if (header.containsKey(HEADER_KEY_MODE)) {
                        withDeliveryMode((int) header.get(HEADER_KEY_MODE));
                        header.remove(HEADER_KEY_MODE);
                    }
                    if (header.containsKey(HEADER_KEY_TYPE)) {
                        withType((String) header.get(HEADER_KEY_TYPE));
                        header.remove(HEADER_KEY_TYPE);
                    }
                    if (header.containsKey(HEADER_KEY_REDELIVERED)) {
                        withRedelivered((boolean) header.get(HEADER_KEY_REDELIVERED));
                        header.remove(HEADER_KEY_REDELIVERED);
                    }
                    if (header.containsKey(HEADER_KEY_PRIORITY)) {
                        withPriority((int) header.get(HEADER_KEY_PRIORITY));
                        header.remove(HEADER_KEY_PRIORITY);
                    }

                    withHeaders(header);
                    withBody(bodyBuffer);

                    return build();

                } catch (final IOException ex) {
                    throw new MessageBuilderException("Unable to create message", ex);
                }

            } else {
                withBody(buffer);
                return build();
            }
        } else {
            withBody(buffer);
            return build();
        }

    }
```

#### Copy headers from Bridge message to JMS message

```java 
package io.nats.bridge.jms.support;

import io.nats.bridge.messages.BaseMessageWithHeaders;
import io.nats.bridge.messages.Message;
import io.nats.bridge.util.FunctionWithException;

import javax.jms.BytesMessage;
import javax.jms.JMSException;
import javax.jms.Session;
import java.util.Map;
import java.util.Set;

public class ConvertBridgeMessageToJmsMessageWithHeaders implements FunctionWithException<Message, javax.jms.Message> {

    private final Session session;

    public ConvertBridgeMessageToJmsMessageWithHeaders(final Session session) {
        this.session = session;
    }

    @Override
    public javax.jms.Message apply(final Message message) throws Exception {
        final BytesMessage bytesMessage = session.createBytesMessage();

        if (message instanceof BaseMessageWithHeaders) {
            copyHeaders(message, bytesMessage);
        }
        bytesMessage.writeBytes(message.getBodyBytes());
        return bytesMessage;
    }

    private void copyHeaders(Message message, BytesMessage bytesMessage) throws JMSException {
        final Map<String, Object> headers = message.headers();
        final Set<String> keys = headers.keySet();
        for (String key : keys) {
            bytesMessage.setObjectProperty(key, headers.get(key));
        }

        if (message.timestamp() != -1)
            bytesMessage.setJMSTimestamp(message.timestamp());

        if (message.expirationTime() != -1)
            bytesMessage.setJMSExpiration(message.expirationTime());

        if (message.deliveryTime() != -1)
            bytesMessage.setJMSDeliveryTime(message.deliveryTime());

        if (message.deliveryMode() != -1)
            bytesMessage.setJMSDeliveryMode(message.deliveryMode());

        if (!Message.NO_TYPE.equals(message.type()))
            bytesMessage.setJMSType(message.type());

        bytesMessage.setJMSRedelivered(message.redelivered());

        if (message.priority() != -1)
            bytesMessage.setJMSPriority(message.priority());

        if (message.correlationID() != null && message.correlationID().trim().length()>0)
            bytesMessage.setJMSCorrelationID(bytesMessage.getJMSCorrelationID());
    }
}

```

####  Copy headers from JMS message to Bridge message

```java 
package io.nats.bridge.jms.support;

import io.nats.bridge.TimeSource;
import io.nats.bridge.jms.JMSMessageBusException;
import io.nats.bridge.messages.Message;
import io.nats.bridge.messages.MessageBuilder;
import io.nats.bridge.util.ExceptionHandler;
import io.nats.bridge.util.FunctionWithException;

import javax.jms.BytesMessage;
import javax.jms.Destination;
import javax.jms.TextMessage;
import java.nio.charset.StandardCharsets;
import java.util.Enumeration;
import java.util.Queue;

public class ConvertJmsMessageToBridgeMessageWithHeaders implements FunctionWithException<javax.jms.Message, Message> {

    private final ExceptionHandler tryHandler;
    private final TimeSource timeSource;
    private final Queue<JMSReply> jmsReplyQueue;

    public ConvertJmsMessageToBridgeMessageWithHeaders(final ExceptionHandler tryHandler, final TimeSource timeSource,
                                                       final Queue<JMSReply> jmsReplyQueue) {
        this.tryHandler = tryHandler;
        this.timeSource = timeSource;
        this.jmsReplyQueue = jmsReplyQueue;
    }

    private void enqueueReply(final long sentTime, final Message reply, final String correlationID, final Destination jmsReplyTo) {
        jmsReplyQueue.add(new JMSReply(sentTime, reply, correlationID, jmsReplyTo));
    }

    private byte[] readBytesFromJMSMessage(final javax.jms.Message jmsMessage) throws Exception {

        if (jmsMessage instanceof BytesMessage) {
            final BytesMessage bytesMessage = (BytesMessage) jmsMessage;
            byte[] buffer = new byte[(int) bytesMessage.getBodyLength()];
            bytesMessage.readBytes(buffer);
            return buffer;
        } else if (jmsMessage instanceof TextMessage) {
            return ((TextMessage) jmsMessage).getText().getBytes(StandardCharsets.UTF_8);
        } else {
            throw new JMSMessageBusException("Unable to read bytes from message " + jmsMessage.getClass().getName());
        }

    }

    @Override
    public Message apply(final javax.jms.Message jmsMessage) throws Exception {

        final Destination jmsReplyTo = jmsMessage.getJMSReplyTo();
        final long startTime = timeSource.getTime();

        byte[] bodyBytes = readBytesFromJMSMessage(jmsMessage);
        final MessageBuilder builder = MessageBuilder.builder().withBody(bodyBytes);
        copyHeaders(builder, jmsMessage);
        if (jmsReplyTo != null) {
            builder.withReplyHandler(reply -> tryHandler.tryWithRethrow(() -> enqueueReply(startTime, reply, jmsMessage.getJMSCorrelationID(), jmsReplyTo), e -> {
                throw new JMSMessageBusException("Unable to send to JMS reply", e);
            }));
        }
        return builder.build();

    }

    private void copyHeaders(final MessageBuilder builder, final javax.jms.Message jmsMessage) throws Exception {


        final Enumeration<String> propertyNames = (Enumeration<String>) jmsMessage.getPropertyNames();
        while (propertyNames.hasMoreElements()) {
            final String propertyName = propertyNames.nextElement();
            if (!propertyName.startsWith("JMS")) //Will make excluding or including configurable with reasonable defaults 
            builder.withHeader(propertyName, jmsMessage.getObjectProperty(propertyName));
        }

        builder.withRedelivered(jmsMessage.getJMSRedelivered());
        builder.withDeliveryTime(jmsMessage.getJMSDeliveryTime());
        builder.withDeliveryMode(jmsMessage.getJMSDeliveryMode());
        builder.withType(jmsMessage.getJMSType());
        builder.withPriority(jmsMessage.getJMSPriority());
        builder.withExpirationTime(jmsMessage.getJMSExpiration());
        builder.withTimestamp(jmsMessage.getJMSTimestamp());
    }

}

```

### CODE REVIEW

**Comment by derekcollison** [comment]
What are the AB/CD?

What hash are you using for integrity for the headers?

File: src/main/java/io/nats/bridge/jms/support/JMSMessageBusBuilder.java [hunk 1]
```diff
@@ -328,8 +403,22 @@ public JMSMessageBusBuilder withSessionCreator(final Function<Connection, Sessio
...
-                getResponseConsumer(), getTimeSource(), getMetrics(), getProducerSupplier(), getConsumerSupplier(), getMetricsProcessor(), getTryHandler(), getJmsBusLogger());
+                getResponseConsumer(), getTimeSource(), getMetrics(), getProducerSupplier(), getConsumerSupplier(),
+                getMetricsProcessor(), getTryHandler(), getJmsBusLogger(), getJmsReplyQueue(), getJmsMessageConverter(), getBridgeMessageConverter());
     }
 
 
+    public Hashtable<String, Object> getJndiProperties() {
+        if (jndiProperties.size() == 0) {
+            jndiProperties.put("java.naming.factory.initial", (String) System.getenv().getOrDefault("NATS_BRIDGE_JMS_NAMING_FACTORY", "org.apache.activemq.artemis.jndi.ActiveMQInitialContextFactory"));
+            jndiProperties.put("connectionFactory.ConnectionFactory", (String) System.getenv().getOrDefault("NATS_BRIDGE_JMS_CONNECTION_FACTORY", "tcp://localhost:61616"));
```
Comment (ColinSullivan1) [inline src/main/java/io/nats/bridge/jms/support/JMSMessageBusBuilder.java:414]: There was a requirement to support the MQSERVER, MQCHLLB, and MQCHLTAB variables.  Would these fall into a custom connection factory?

**Comment by ColinSullivan1** [comment]
Just curious... In the protocol why the last two markers?
```java
           streamOut.writeByte(Protocol.MARKER_AB);
           streamOut.writeByte(Protocol.MARKER_CD);
```
Is that to keep the fingerprint(s) + size + hash to 8 bytes (register sized)?  If for sizing only, they could be reserved instead.

**Comment by RichardHightower** [comment]
```
Rick Hightower  8:38 AM
BTW… There is a lot of code to review for that PR so I pulled the important parts into the README for the PR .. along with descriptions of the wire protocol. I’d skip the code and skim the README.






Derek Collison  10:06 AM
ok will do, just want to make sure we coordinate on the encoding for the header properties etc since we may want to extend (not client proto) to other plain NATS clients..,

Rick Hightower  10:14 AM
ok
10:15
I versioned the encoding, so we can change it.
10:15
image.png 
image.png

```

image

image

image


```
10:16
image.png 
image.png


10:17
Then it does a quick check to see if headers exists and if not … it just just a bytes…
10:17
image.png 
image.png
```

image

```
10:18
Put two hash checks in there … mainly so someone does not accidentally send bad bytes.

Derek Collison  10:25 AM
Yea I would have maybe put a single magic byte at the beginning to denote that headers were even present, so plain byte bags worked but the check was really simple. Then they could have the versions etc but I think we would only support KEY : VALUE where both are strings (UTF8)..

10:26
But the AB marker could be the check which is exactly what you do anyway, so should work fine.. Its 6 bytes though, might be better to make it 4 or 8..

Rick Hightower  10:26 AM
ok

Derek Collison  10:26 AM
I would say 4. So START_ MAGIC_BYTE | MAJOR_VERSION_BYTE | MINOR_VERSION_BYTE | END_MAGIC_BYTE

```

```
10:27
Need to think a little bit on what they are too as to not have false collisions with normal binary payloads etc..

Rick Hightower  10:29 AM
`[0xde] [0xad] [0xbe] [0xef] [Major] [Minor] [0xab] [0xcd] `(8 bytes deadbeef major minor abcd) ? Like that?
10:30
or `[0xab][Major] [Minor][0xcd]` for four byte …. 8 seems safer.. less likely to collide.

Derek Collison  10:31 AM
lol I did deadbeef and cafebabe back in the day..

(Colin mentioned that to me which I why I added it. 

```

```
1 reply
Today at 10:33 AMView thread

Derek Collison  10:31 AM
If the START_MAGIC is below normal ASCII and is not another magic for payloads like PNG, JPG, etc we should be ok.
10:32
I like 4 just because I envision Billions of these being sent and the BW waste :slightly_smiling_face:

Rick Hightower  10:32 AM
and the hash/length and hash/length should avoid most collisions .. we can revert back to bytes if they don’t match.

Derek Collison  10:33 AM
Which hash you using? For JetStream I upped it big time and am using a highway hash, 10Gb per core at speed based on impl, but well beyond crcs etc..
10:33
All JS messages are HHashed with time, subject and msg payload..

Rick Hightower  10:33 AM
similar hash as java  string
10:34
image.png 
image.png

```

image



```

Derek Collison  10:34 AM
for driving maps in Java etc?
10:34
That will have collisions..
10:34
Will be fast but..

Rick Hightower  10:36 AM
so I covert header to JSON then JSON to bytes and then hash it. then i hash the payload… i write length/hash, length/hash for both headers and payload.. so two 1 in 64 billion chances of a collision. I will take those odds.

Derek Collison  10:38 AM
So headers are JSON? I would want to discuss, JSON is good but expensive, and for simple K:V encodings I think something simple and custom would be way better. Even if we say keys and values are strings with no spaces, and use spaces for the delimiter in the encoding.
10:38
So `KEY VALUE KEY2 VALUE2`

Rick Hightower  10:38 AM
I was thinking easy to do.. but I can of course encode headers. :slightly_smiling_face:

Derek Collison  10:39 AM
I like JSON for slow path stuff, but anything that can be highspeed care deeply about encoding and avoiding double encoding etc..

Rick Hightower  10:39 AM
so KEY TYPE VALUE where type for JMS is (int, double, float, long, string)

Derek Collison  10:39 AM
So JMS can have non string values?

2 replies
Last reply today at 10:41 AMView thread

Derek Collison  10:40 AM
Been so long I can't remember :wink:
10:40
Maybe our header has a type too, one for K/V all strings, and one for typed values like JMS needs? WDYT?
10:40
That make sense?

Rick Hightower  10:41 AM
LEN_SHORT(2) <STRING_BYTES>(N) TYPE_BYTE (1) ENCODED_VALUE(1 to 8 bytes) or STRING

Derek Collison  10:42 AM
That works I think, a bit more complicated than just string string.. Reason I am thinking that way is we have 40+ client language impls that would need to code this up, most are maintained not by us.. String string with space for delimiter is trivial..
:+1:
1


Rick Hightower  10:43 AM
JMS message btw
10:43
    boolean getBooleanProperty(String name) throws JMSException;
    byte getByteProperty(String name) throws JMSException;
    short getShortProperty(String name) throws JMSException;
    int getIntProperty(String name) throws JMSException;
    long getLongProperty(String name) throws JMSException;
    float getFloatProperty(String name) throws JMSException;
    double getDoubleProperty(String name) throws JMSException;
    String getStringProperty(String name) throws JMSException;
10:44
The issues with string/string for JMS is I need the types to copy the headers.

Derek Collison  10:44 AM
ok gotcha..
10:44
Let's do your suggestion : LEN_SHORT(2) <STRING_BYTES>(N) TYPE_BYTE (1) ENCODED_VALUE(1 to 8 bytes) or STRING
:+1:
1

10:45
vs json and we can plumb that into our clients and help others, or they can copy some of our code..

Rick Hightower  10:45 AM
    void setBooleanProperty(String name, boolean value) throws JMSException;
    void setByteProperty(String name, byte value) throws JMSException;
    void setShortProperty(String name, short value) throws JMSException;
    void setIntProperty(String name, int value) throws JMSException;
    void setLongProperty(String name, long value) throws JMSException;
    void setFloatProperty(String name, float value) throws JMSException;
    void setDoubleProperty(String name, double value) throws JMSException;
    void setStringProperty(String name, String value) throws JMSException;
    void setObjectProperty(String name, Object value) throws JMSException;
10:45
ok
10:45
cool

Derek Collison  10:45 AM
thx!

Rick Hightower  10:45 AM
btw setObjectProperty does not do any conversion
10:45
can I cut/paste this convo for the PR comments?

Derek Collison  10:45 AM
How is it encoded on the wire?
10:45
Yes of course..

Rick Hightower  10:47 AM
JMS? Encoding is not specified. It is not like AMQP or MQTT or WebSocket or HTTP. JMS defines an interface as JSR but that is just an API. Everyone implementation can vary the wire protocol. (edited) 

Derek Collison  10:47 AM
How will you do Object?
10:47
It has to have a way to marshal itself to bytes no?

Rick Hightower  10:48 AM
Object can only be String, Integer, Boolean, Long, Short, Byte,
New

Derek Collison  10:49 AM
:thumbsup::skin-tone-2:

Rick Hightower  10:49 AM
so primitive wrappers and String.
10:49
Very limited.
10:50
Thank you Derek.
```

**Comment by RichardHightower** [comment]
@ColinSullivan1 it was a byte marker to identify the byte array

File: src/main/java/io/nats/bridge/jms/support/JMSMessageBusBuilder.java (hunk 1 above)
Comment (RichardHightower) [reply src/main/java/io/nats/bridge/jms/support/JMSMessageBusBuilder.java:414]: environment vars? We will have a config file.

**Comment by RichardHightower** [comment]
https://github.com/nats-io/nats-jms-mq-bridge/pull/41 I am closing this one w/o a merge.

**Comment by RichardHightower** [comment]
Since we went a different way, I added a new PR. 
https://github.com/nats-io/nats-jms-mq-bridge/pull/41

**Pull request closed without merging**